File: backend/src/financial_simulator/app/server/routers/schedules/schedules.py
import logging
from typing import Union

from fastapi import APIRouter
from sqlalchemy import ColumnElement
from sqlalchemy.orm import InstrumentedAttribute

from financial_simulator.app.database.schema import (
    ScheduleType,
    Schedule,
)

from financial_simulator.app.server.util.typed_collection import TypedCollection
from .all_schedule import AllScheduleGet, AllSchedulePost, all_schedule_model_mapper
from .any_schedule import AnyScheduleGet, AnySchedulePost, any_schedule_model_mapper
from .daily_schedule import (
    daily_schedule_model_mapper,
    DailyScheduleGet,
    DailySchedulePost,
)
from .day_schedule import (
    day_schedule_model_mapper,
    DayScheduleGet,
    DaySchedulePost,
)
from .from_schedule import (
    from_schedule_model_mapper,
    FromScheduleGet,
    FromSchedulePost,
)
from .monthly_schedule import (
    MonthlyScheduleGet,
    MonthlySchedulePost,
    monthly_schedule_model_mapper,
)
from .until_schedule import (
    until_schedule_model_mapper,
    UntilScheduleGet,
    UntilSchedulePost,
)
from .range_schedule import (
    range_schedule_model_mapper,
    RangeScheduleGet,
    RangeSchedulePost,
)
from .weekly_schedule import (
    WeeklyScheduleGet,
    WeeklySchedulePost,
    weekly_schedule_model_mapper,
)
from .yearly_schedule import (
    YearlyScheduleGet,
    YearlySchedulePost,
    yearly_schedule_model_mapper,
)
from ...util.query_params import DefaultQueryParams

logger = logging.getLogger(__name__)

router = APIRouter(
    prefix="/schedules",
    tags=["schedules"],
)


class ScheduleQueryParams(DefaultQueryParams):
    type: ScheduleType | None = None

    def query_order_by(self) -> Union[InstrumentedAttribute[str], None]:
        return Schedule.name

    def query_where(self) -> Union[ColumnElement[bool], None]:
        if self.type is None:
            return None
        return Schedule.type == self.type


TypedCollection(
    table_model=Schedule,
    query_params_class=ScheduleQueryParams,
    get_model=Union[
        DailyScheduleGet,
        DayScheduleGet,
        YearlyScheduleGet,
        MonthlyScheduleGet,
        WeeklyScheduleGet,
        FromScheduleGet,
        UntilScheduleGet,
        RangeScheduleGet,
        AllScheduleGet,
        AnyScheduleGet,
    ],
    post_model=Union[
        DailySchedulePost,
        DaySchedulePost,
        YearlySchedulePost,
        MonthlySchedulePost,
        WeeklySchedulePost,
        FromSchedulePost,
        UntilSchedulePost,
        RangeSchedulePost,
        AllSchedulePost,
        AnySchedulePost,
    ],
    model_mappers={
        ScheduleType.DAILY: daily_schedule_model_mapper,
        ScheduleType.DAY: day_schedule_model_mapper,
        ScheduleType.YEARLY: yearly_schedule_model_mapper,
        ScheduleType.MONTHLY: monthly_schedule_model_mapper,
        ScheduleType.WEEKLY: weekly_schedule_model_mapper,
        ScheduleType.FROM: from_schedule_model_mapper,
        ScheduleType.UNTIL: until_schedule_model_mapper,
        ScheduleType.RANGE: range_schedule_model_mapper,
        ScheduleType.ALL: all_schedule_model_mapper,
        ScheduleType.ANY: any_schedule_model_mapper,
    },
).add_endpoints(
    router=router,
)

File: frontend/app/page_params/schedules.ts
import {PageParams} from "./PageParams";
import type {
    AllScheduleGet,
    AnyScheduleGet,
    DailyScheduleGet,
    DayScheduleGet,
    FromScheduleGet,
    MonthlyScheduleGet,
    RangeScheduleGet,
    UntilScheduleGet,
    WeeklyScheduleGet,
    YearlyScheduleGet
} from "../../client";
import type {NavbarLinkTree} from "~/components/layout/NavbarLink";

export const SCHEDULES_HREF = '/schedule';
export const SCHEDULES_PAGE_DESCRIPTION = 'Schedules';

export const DAILY_SCHEDULE_PARAMS = new PageParams<DailyScheduleGet>(
    "Daily schedules",
    SCHEDULES_HREF + "/daily",
    "daily schedule",
    "Daily Schedule",
    "name",
    (item: DailyScheduleGet) => ({
        id: item.id,
        name: item.name,
    }),
)

export const DAY_SCHEDULE_PARAMS = new PageParams<DayScheduleGet>(
    "Day schedules",
    SCHEDULES_HREF + "/day",
    "day schedule",
    "Day Schedule",
    "name",
    (item: DayScheduleGet) => ({
        id: item.id,
        name: item.name,
    }),
)

export const WEEKLY_SCHEDULE_PARAMS = new PageParams<WeeklyScheduleGet>(
    "Weekly schedules",
    SCHEDULES_HREF + "/weekly",
    "weekly schedule",
    "Weekly Schedule",
    "name",
    (item: WeeklyScheduleGet) => ({
        id: item.id,
        name: item.name,
    }),
)

export const MONTHLY_SCHEDULE_PARAMS = new PageParams<MonthlyScheduleGet>(
    "Monthly schedules",
    SCHEDULES_HREF + "/monthly",
    "monthly schedule",
    "Monthly Schedule",
    "name",
    (item: MonthlyScheduleGet) => ({
        id: item.id,
        name: item.name,
    }),
)

export const YEARLY_SCHEDULE_PARAMS = new PageParams<YearlyScheduleGet>(
    "Yearly schedules",
    SCHEDULES_HREF + "/yearly",
    "yearly schedule",
    "Yearly Schedule",
    "name",
    (item: YearlyScheduleGet) => ({
        id: item.id,
        name: item.name,
    }),
)

export const FROM_SCHEDULE_PARAMS = new PageParams<FromScheduleGet>(
    "From schedules",
    SCHEDULES_HREF + "/from",
    "from schedule",
    "From Schedule",
    "name",
    (item: FromScheduleGet) => ({
        id: item.id,
        name: item.name,
    }),
)

export const UNTIL_SCHEDULE_PARAMS = new PageParams<UntilScheduleGet>(
    "Until schedules",
    SCHEDULES_HREF + "/until",
    "until schedule",
    "Until Schedule",
    "name",
    (item: UntilScheduleGet) => ({
        id: item.id,
        name: item.name,
    }),
)

export const RANGE_SCHEDULE_PARAMS = new PageParams<RangeScheduleGet>(
    "Range schedules",
    SCHEDULES_HREF + "/range",
    "range schedule",
    "Range Schedule",
    "name",
    (item: RangeScheduleGet) => ({
        id: item.id,
        name: item.name,
    }),
)

export const ANY_SCHEDULE_PARAMS = new PageParams<AnyScheduleGet>(
    "Any schedules",
    SCHEDULES_HREF + "/any",
    "any schedule",
    "Any Schedule",
    "name",
    (item: AnyScheduleGet) => ({
        id: item.id,
        name: item.name,
    }),
)

export const ALL_SCHEDULE_PARAMS = new PageParams<AllScheduleGet>(
    "All schedules",
    SCHEDULES_HREF + "/all",
    "all schedule",
    "All Schedule",
    "name",
    (item: AllScheduleGet) => ({
        id: item.id,
        name: item.name,
    }),
)

export const SCHEDULES_NAVBAR_LINK_TREE: NavbarLinkTree = {
    navbarLinkHref: SCHEDULES_HREF,
    navbarLinkLabel: SCHEDULES_PAGE_DESCRIPTION,
    navbarLinkChildren: [
        DAILY_SCHEDULE_PARAMS,
        DAY_SCHEDULE_PARAMS,
        WEEKLY_SCHEDULE_PARAMS,
        MONTHLY_SCHEDULE_PARAMS,
        YEARLY_SCHEDULE_PARAMS,
        FROM_SCHEDULE_PARAMS,
        UNTIL_SCHEDULE_PARAMS,
        RANGE_SCHEDULE_PARAMS,
        ALL_SCHEDULE_PARAMS,
        ANY_SCHEDULE_PARAMS,
    ]
}
